packages/block-editor/src/components/list-view/use-list-view-client-ids.js

```diff
@@ -16,14 +17,30 @@ export default function useListViewClientIds( { blocks, rootClientId } ) {
 				getDraggedBlockClientIds,
 				getSelectedBlockClientIds,
 				__unstableGetClientIdsTree,
-			} = select( blockEditorStore );
+				getBlockEditingMode,
+			} = unlock( select( blockEditorStore ) );
+
+			const removeDisabledBlocks = ( tree ) => {
+				return tree.flatMap( ( { clientId, innerBlocks, ...rest } ) => {
+					if ( getBlockEditingMode( clientId ) === 'disabled' ) {
+						return removeDisabledBlocks( innerBlocks );
+					}
+					return [
+						{
+							clientId,
+							innerBlocks: removeDisabledBlocks( innerBlocks ),
+							...rest,
+						},
+					];
+				} );
+			};
```

Comment:
This will return a new array reference on each call and cause the whole ListView to be rendered, even if the client IDs tree is the same.

Should we extract this into a new private memoized selector?

Reply:
I tested something out by adding the following selector to packages/block-editor/src/store/private-selectors.js.


```js
const EMPTY_ARRAY = [];
export const getEnabledBlocks = createSelector(
	( state, blocks = EMPTY_ARRAY ) =>
		blocks.flatMap( ( { clientId, innerBlocks, ...rest } ) => {
			if ( getBlockEditingMode( state, clientId ) === 'disabled' ) {
				return getEnabledBlocks( state, innerBlocks );
			}
			return [
				{
					clientId,
					innerBlocks: getEnabledBlocks( state, innerBlocks ),
					...rest,
				},
			];
		} ),
	( state ) => [ state.blockEditingModes, state.blocks.parents ]
);
```

Used in  packages/block-editor/src/components/list-view/use-list-view-client-ids.js

```js
...
			const {
				getDraggedBlockClientIds,
				getSelectedBlockClientIds,
				__unstableGetClientIdsTree,
				getEnabledBlocks,
			} = unlock( select( blockEditorStore ) );
...
				clientIdsTree: getEnabledBlocks(
					blocks ?? __unstableGetClientIdsTree( rootClientId )
				),
...
```

And it does prevent entire tree rerenders, which might be helpful for posts with lots of blocks:

## Before

https://github.com/WordPress/gutenberg/assets/6458278/eadef3c4-76d1-4ffc-9510-1b13ce264aee



## After

https://github.com/WordPress/gutenberg/assets/6458278/40e154ae-1aa6-45cb-b739-115c17195b94

I'm just not sure about the dependency array `[ state.blockEditingModes, state.blocks.parents ]`, which I copied from `getExplicitBlockEditingMode()` since that function is called in `getEnabledBlocks()` anyway.